Enable websocket support

nareddyt opened the pull request:
Fixes #77 

## Description

Allows proxy to handle websocket connections. 

### Major changes:

- Fix hardcoded HTTP protocol version. Hardcoded to HTTP2, which does not support websockets (at least not well)
- Prevent websocket connections from being proxied through a `timeoutHandler`, as websockets are long-lived connections and should not have timeouts enforced on them.
- Added configuration variable to opt-in to websocket support.

### Other Notes

This is an opt-in feature because it can be potentially abused (see discussion in this PR). By opting-in, the user acknowledges this risk.

**Checklist**:
- [x] updated docs
- [x] unit tests added
- [x] related issues referenced
- [x] ready for review

desimone commented:
My first thought was to do something like [this](https://github.com/pomerium/pomerium/blob/7487de94df7c0c585ea8123bf7c1358eaa98fde3/internal/https/https.go#L196-L207) and force downgrade (or use the appropriate fancywriter) as middleware. 

[Relevant comment](https://github.com/golang/go/issues/14797#issuecomment-196103814) (I first thought we could just implement a hijacker for the http2FancyWriter) about hijacker and http2 being incompatible.

desimone commented on internal/log/middleware.go:
```diff
@@ -164,7 +164,7 @@ func AccessHandler(f func(r *http.Request, status, size int, duration time.Durat
 		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 			start := time.Now()
-			lw := NewWrapResponseWriter(w, 2)
+			lw := NewWrapResponseWriter(w, 1)
```
```go 
// The protocol version for incoming server requests.
//
// For client requests, these fields are ignored. The HTTP
// client code always uses either HTTP/1.1 or HTTP/2.
// See the docs on Transport for details.
Proto      string // "HTTP/1.0"
ProtoMajor int    // 1
ProtoMinor int    // 0
```
```go
lw := NewWrapResponseWriter(w, r.ProtoMajor)
````

🤷‍♂ might work? Thoughts?

nareddyt replied:
Oh maybe, let me try that. I'll also play around with that you said in the above comment.

nareddyt commented:
Your suggestion works! I tested it with Home Assistant and Jupyter backends, taking note of the protocol versions in chrome dev tools. All http2 requests are being handled correctly, and all websocket requests are being hijacked as http1.1 requests.

I also added simple switch like you suggested. Now the proxy does not use the `TimeoutHandler` for websocket connections, but defaults to using it for all other requests.

I updated the PR with this code. Could you take a look? Let me know if I need to re-factor anything as well. I will work on adding tests later.

desimone commented:
Glad it worked! Code looks good. 

The only thing I can think about right now is that withou timeout handler on websocket requests we could run into a familiar issue as #40 . It's not a hair on fire concern as that assumes very specific attack conditions but it's something we should try to guard against as much as possible with these long-lived non-timeoutable requests. I'll think about mitigations more -- both for this PR and in general... but if you can think of any solutions that would be very welcome.

desimone commented on proxy/handlers.go:
```diff
@@ -501,3 +501,18 @@ func (p *Proxy) GetSignOutURL(authenticateURL, redirectURL *url.URL) *url.URL {
+		cn := r.Header.Get("Connection")
+		up := r.Header.Get("Upgrade")
+
+		if r.ProtoMajor == 1 && strings.Contains(cn, "Upgrade") && strings.Contains(up, "websocket") {
```
nit: Although the line gets longer, you can probably inline the r.Get()'s which saves the call if r.Proto is not 1.

desimone replied:
Should we match for exact string for websocket and upgrade vs contains?

nareddyt replied:
Good points, I can make both changes tomorrow.

nareddyt commented:
Yeah, we can't add a timeout to websockets because they are long-lived connections by definition. This is a fundamental issue with websockets and hard to solve...
- https://security.stackexchange.com/questions/48378/anti-dos-websockets-best-practices
- https://samsaffron.com/archive/2015/12/29/websockets-caution-required

Here are some possible mitigations that are not timeout based and implementable in pomerium (in my opinion):
- **Opt-in Websocket Support Config Flag:** Make websocket support opt-in via the configuration variables. Users must explicitly acknowledge the possibility of abuse to enable websocket support.
- **Limit Number of Websocket Connections:** Set hard limits on the number of active websocket connections. Make threshold editable via the configuration. Limits can be enforced per source IP (does not prevent against distributed attacks) or globally across all source IPs (incorrect threshold can cause availability issues).

desimone commented:
Option 1 feels like it strikes the right balance to me. Thoughts?

nareddyt commented:
I can implement option 1 in this PR. We can look into the possibility of rate limiting techniques later 👍

nareddyt commented:
@desimone ready for review!

desimone commented:
@nareddyt Looks great, thank you!
